### PR TITLE
Consider Scrollable location in text selection drag events

### LINKED ISSUE #102484: TextField does not scroll when selecting text
## Steps to Reproduce

1. Execute the code sample
2. Enter a large text to the text field (e.g. copy-paste code sample to it)
3. Try to select the entire text

**Expected results:** 
Area inside TextField should scroll when mouse rich to the edge

**Actual results:** 
Area inside TextField is not scrolling

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: const EditorPage(title: 'Editor Demo'),
    );
  }
}

class EditorPage extends StatefulWidget {
  const EditorPage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<EditorPage> createState() => _EditorPageState();
}

class _EditorPageState extends State<EditorPage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(widget.title)),
      body: SingleChildScrollView(
        child: Column(
          children: const [
            TextField(
              style: TextStyle(fontSize: 24),
              decoration: InputDecoration(
                border: InputBorder.none,
                hintText: 'Please enter title',
                contentPadding: EdgeInsets.symmetric(horizontal: 16),
              ),
            ),
            TextField(
              maxLines: null,
              style: TextStyle(fontSize: 18, fontWeight: FontWeight.normal),
              decoration: InputDecoration(
                border: InputBorder.none,
                hintText: 'Please enter content',
                isCollapsed: true,
                contentPadding: EdgeInsets.symmetric(horizontal: 16),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

<details>
  <summary>Logs</summary>

```
```

```
```

```
flutter doctor -v

[✓] Flutter (Channel beta, 2.13.0-0.2.pre, on macOS 12.3.1 21E258 darwin-arm, locale zh-Hans-CN)
    • Flutter version 2.13.0-0.2.pre at /Users/feimeng/develop/flutter/flutter-beta
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 8662e22bac (5 days ago), 2022-04-20 08:21:52 -0700
    • Engine revision 24a02fa5ee
    • Dart version 2.17.0 (build 2.17.0-266.5.beta)
    • DevTools version 2.12.2
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/feimeng/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.66.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 12.3.1 21E258 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 100.0.4896.88

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

https://user-images.githubusercontent.com/12608693/165046328-8d311910-9c53-4062-ac4f-12952d87ad37.mp4

### LINKED ISSUE #104500: Implementation of `EditableTextState._scheduleShowCaretOnScreen`
Currently it's not possible to select more text than it's visible in a scrollable. Because the function is always trying to show a padded caret rect, which for selected text happens to be at the beginning of the selection. This is because `_FloatingCursorPainter.paint` skips painting for non-collapsed selection and updating the cashed caret rect is performed at the end of the painting via callback to `RenderEditable`. 

Visually there is no caret on non-collapsed selection, then what to show? Why should the beginning of the selection be treated as the caret position and not the end? Maybe the function should have another name and `EditableTextState._currentCaretRect` be updated to `null` . Or `_currentCaretRect` should be updated even for non-collapsed selection and placed at the extent of the selection.

Probably the simplest solution would be to add 
```dart
final Rect rectToReveal;
final TextSelection selection = textEditingValue.selection;
if (selection.isCollapsed) {
  rectToReveal = targetOffset.rect;
} else {
  final List<Rect> selectionBoxes = renderEditable.getBoxesForSelection(selection);
  rectToReveal = selection.baseOffset < selection.extentOffset ?
    selectionBoxes.last : selectionBoxes.first;
}
```
and change 
```dart
rect: caretPadding.inflateRect(targetOffset.rect)
```
to
```dart
rect: caretPadding.inflateRect(rectToReveal)
```

Also note there is another issue with scrolling while selecting text by dragging the mouse, but that has to do with the implementation of `TextSelectionGestureDetectorBuilder.onDragSelectionUpdate`.

### PR DESCRIPTION
**Pull request by justinmc**
Previously, edge scrolling was incorrect if it caused a Scrollable to scroll.  Our math only considered scrolling the RenderEditable itself, but not a containing Scrollable.  This PR fixes this by taking both into consideration.

| Old | New |
| --- | --- |
| out | out |

## The problem
Drag gestures in GestureDetector don't consider scrolling that happens during the event, as demonstrated in [this dartpad](https://dartpad.dartlang.org/?id=8b64cae32e4fabe7208036996f71d6ca).  Because of this, we have to compensate for the scrolling manually when calculating the selection area during edge scrolling.

The scrolling location is not manually calculated, it's just done via [renderEditable.showOnScreen](https://github.com/flutter/flutter/blob/d5ac8eea4ed4e18df22f24ebeffac2910302a02c/packages/flutter/lib/src/widgets/editable_text.dart#L2936), so we don't have the calculated scroll position anywhere.

Previously, this compensation was done only for scrolling that happens inside the RenderEditable ([here](https://github.com/flutter/flutter/blob/d5ac8eea4ed4e18df22f24ebeffac2910302a02c/packages/flutter/lib/src/widgets/text_selection.dart#L1732)).  However, it's also possible that a tall text input in a small Scrollable causes the Scrollable to scroll ([example in the issue](https://github.com/flutter/flutter/issues/102484)), not the RenderEditable, and this must be compensated for, too.

## The current solution
Grab the scroll position using `Scrollable.of(context).position.pixels` and use it to compensate for any scrolling when calculating the selection Rect.

## Related

Fixes https://github.com/flutter/flutter/issues/102484
Partially fixes https://github.com/flutter/flutter/issues/104500 (desktop only fix)

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
I know passing around BuildContext isn't ideal.. looking a bit more into this, it looks like you could do this another way. 

Let me know if I am off here, but it looks like the Scrollable in EditableTextState is the one you are after. Instead of looking it up using a context, it may be cleaner to refactor the ScrollController logic of that Scrollable.

If you were able to access the scroll controller, you could access the position similarly without a build context.

https://github.com/flutter/flutter/blob/896e5b332d364cbe407966787115ce6cc4e6dde4/packages/flutter/lib/src/widgets/editable_text.dart#L1584

I am not super familiar with editable text, but from what I can tell, a user-provided scroll controller is passed down from the higher level classes like TextField.scrollController and CupertinoTextField.scrollController. If the user has not provided a ScrollController, one is created all the way down in EditableTextState as a fallback.

If instead, you were to take the fallback creation out of EditableTextState and instead create the fallback controller in the higher level classes, the controller (whether it is user provided or the fallback) can be passed to the *TextSelectionGestureDetectorBuilders and then in turn passed further down to EditableTextState. That would remove the need to call Scrollable.of, and no passing around a context that could become invalid.

What do you think? Let me know if you have questions or if I've been unclear. :)

**Comment by justinmc** [comment]
@Piinks Actually I explicitly need the Scrollable above the EditableText, not inside of it.  Previously we were compensating only for scrolling inside of the EditableText (done [here](https://github.com/flutter/flutter/blob/d5ac8eea4ed4e18df22f24ebeffac2910302a02c/packages/flutter/lib/src/widgets/text_selection.dart#L1733) with renderEditable.offset).  This PR aims to compensate for any scrolling happening above the EditableText in the widget tree.  The [issue](https://github.com/flutter/flutter/issues/102484) has an example.

I'll see if I can do this in a cleaner way at all.  I'll tag you for re-review.

**Comment by Piinks** [comment]
Oh! Ok, sorry I did not realize. :) So there may, or may not, be an enclosing Scrollable above the TextField. I think in this case then, `Scrollable.of` if the best approach. Since the Scrollable is outside of the TextField, there is not another way for it to know about it.

**Comment by Hixie** [comment]
Hey @justinmc, is this PR still on your radar?

**Comment by justinmc** [comment]
Ah I had forgotten about it, will bring it back to life.

**Comment by justinmc** [comment]
@Piinks This is ready for review again.  Sorry for the long delay since you last reviewed, but I've now used the `Scrollable.of` approach and cleaned up the PR.

File: packages/flutter/test/widgets/text_selection_test.dart [hunk 1]
```diff
@@ -1329,6 +1330,160 @@ void main() {
       });
     });
   });
+
+  testWidgets('Mouse edge scrolling works in an outer scrollable', (WidgetTester tester) async {
+    // Regression test for https://github.com/flutter/flutter/issues/102484
+    final TextEditingController controller = TextEditingController(
+      text: 'I love flutter!\n' * 8,
```
Comment (Piinks) [inline packages/flutter/test/widgets/text_selection_test.dart:1337]: Best test content ever. 😜

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 2]
```diff
@@ -2134,12 +2147,22 @@ class TextSelectionGestureDetectorBuilder {
...
           ? Offset(renderEditable.offset.pixels - _dragStartViewportOffset, 0.0)
           : Offset(0.0, renderEditable.offset.pixels - _dragStartViewportOffset);
-
+      final ScrollableState? scrollableState =
+          delegate.editableTextKey.currentContext == null
+              ? null
+              : Scrollable.of(delegate.editableTextKey.currentContext!);
+      final double currentScrollDy = scrollableState == null
+          ? 0.0
+          : scrollableState.position.pixels;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/text_selection.dart:2159]: Can this and the same chunk of lines starting at 2122 be wrapped up in a method?

File: packages/flutter/test/widgets/text_selection_test.dart [hunk 3]
```diff
@@ -79,8 +79,9 @@ void main() {
       forcePressEnabled: forcePressEnabled,
       selectionEnabled: selectionEnabled,
     );
+
     final TextSelectionGestureDetectorBuilder provider =
-    TextSelectionGestureDetectorBuilder(delegate: delegate);
+        TextSelectionGestureDetectorBuilder(delegate: delegate);
```
Comment (Piinks) [inline packages/flutter/test/widgets/text_selection_test.dart:84]: Nit
```suggestion
      TextSelectionGestureDetectorBuilder(delegate: delegate);
```

**Review by Piinks — COMMENTED** [review]
> Hopefully will also fix https://github.com/flutter/flutter/issues/104500

Does the rework fix this now? It looks like the last update to OP and the issue was May 27

**Comment by justinmc** [comment]
I tried it and this seems to only fully fix https://github.com/flutter/flutter/issues/104500 on desktop.  I'll update that issue.

**Review by Piinks — APPROVED** [review]
Thanks for checking on the other issue. This LGTM!

**Pull request merged**